source/sigv4_quicksort.c

```diff
@@ -0,0 +1,171 @@
+                  size_t itemSize );
+
+/**
+ * @brief A helper function to perform quicksort on a subarray.
+ *
+ * @param[in] pArray The array to be sorted.
+ * @param[in] low The low index of the array.
+ * @param[in] high The high index of the array.
+ * @param[in] itemSize The amount of memory per entry in the array.
+ * @param[out] comparator The comparison function to determine if one item is less than another.
+ */
+static void quickSortHelper( void * pArray,
+                             size_t low,
+                             size_t high,
+                             size_t itemSize,
+                             ComparisonFunc_t comparator );
+
+/**
+ * @brief A helper function to partition a subarray using the last element
+ * of the array as the pivot. All items smaller than the pivot end up
+ * at its left while all items greater than end up at its right.
+ *
+ * @param[in] pArray The array to be sorted.
+ * @param[in] low The low index of the array.
+ * @param[in] high The high index of the array.
+ * @param[in] itemSize The amount of memory per entry in the array.
+ * @param[out] comparator The comparison function to determine if one item is less than another.
+ *
+ * @return The index of the pivot
+ */
+static size_t partition( void * pArray,
+                         size_t low,
+                         size_t high,
+                         size_t itemSize,
+                         ComparisonFunc_t comparator );
+
+/*-----------------------------------------------------------*/
+
+static void swap( void * pFirstItem,
+                  void * pSecondItem,
+                  size_t itemSize )
+{
+    uint8_t * pFirstByte = pFirstItem;
+    uint8_t * pSecondByte = pSecondItem;
+
+    if( ( pFirstItem != NULL ) && ( pSecondItem != NULL ) )
+    {
+        /* Swap one byte at a time. */
+        while( itemSize-- )
+        {
+            uint8_t tmp = *pFirstByte;
+            *pFirstByte++ = *pSecondByte;
+            *pSecondByte++ = tmp;
+        }
+    }
+}
+
+static void quickSortHelper( void * pArray,
+                             size_t low,
+                             size_t high,
+                             size_t itemSize,
+                             ComparisonFunc_t comparator )
+{
+    size_t stack[ high - low + 1 ];
+    /* Low and high are first two items on the stack. */
+    size_t top = 0;
+
+    stack[ top++ ] = low;
+    stack[ top++ ] = high;
+
+    while( top > 0 )
+    {
+        size_t partitionIndex;
+        high = stack[ --top ];
+        low = stack[ --top ];
+
+        partitionIndex = partition( pArray, low, high, itemSize, comparator );
+
+        if( ( partitionIndex != 0U ) && ( partitionIndex - 1U > low ) )
+        {
+            stack[ top++ ] = low;
+            stack[ top++ ] = partitionIndex - 1U;
+        }
+
+        if( partitionIndex + 1U < high )
+        {
+            stack[ top++ ] = partitionIndex + 1U;
+            stack[ top++ ] = high;
```

Comment:
Consider pushing the larger of the two partitions onto the stack first, as glibc does. This reduces the space complexity from O(n) to O(log n). You could use a 32 element stack for arrays of length up to 4 billion.

For example
```suggestion
        size_t lo1, lo2, hi1, hi2;
        size_t len1, len2;
        len1 = ( ( partitionIndex != 0U ) && ( partitionIndex - 1U > low ) ) ? partitionIndex - 1U - low : 0U;
        len2 = ( partitionIndex + 1U < high ) ? high - partitionIndex - 1U : 0U;

        if( len1 >= len2 )
        {
            lo1 = low;
            hi1 = partitionIndex - 1U;
            lo2 = partitionIndex + 1U;
            hi2 = high;
        }
        else
        {
            lo1 = partitionIndex + 1U;
            hi1 = high;
            lo2 = low;
            hi2 = partitionIndex - 1U;
            /* Also swap the lengths so len1 > len2. */
            len1 ^= len2;
            len2 ^= len1;
            len1 ^= len2;
        }

        if( len1 > 0U )
        {
            stack[ top++ ] = lo1;
            stack[ top++ ] = hi1;
        }
        if( len2 > 0U )
        {
            stack[ top++ ] = lo2;
            stack[ top++ ] = hi2;
```

Reply:
Unfortunately, I still don't think there's any way around actually allocating the number of entries needed for the stack. Because C90 forbids variable-length arrays, we need to just use the macro defined in the config as the length of the stack. I've used your code though in case we'd like to use `malloc` in the future.

Reply:
Makes sense to me if you want to optimize the stack as much as possible. You could also just assume the worst case and use `8 * sizeof( size_t )`